server.js

```diff
@@ -15,6 +17,121 @@ app.get("/", function (request, response) {
+  const datequery = request.query.date
+  const term = request.query.term
+  const date = moment(datequery, 'YYYY-MM-DD', true)
+
+  if (moment(date).isValid()) {
+    let result = bookings.filter(booking => moment(datequery).isBetween(booking.checkInDate, booking.checkOutDate) || moment(date).isSame(booking.checkInDate) || moment(date).isSame(booking.checkOutDate))
+    result.length > 0 ? response.send(result) : response.send('Not found a booking on this date')
+  } else if (datequery && !moment(date).isValid()) {
+    response.status(400).json('Please enter a valid date, format should like "YYYY-MM-DD"')
+  }
+  if (term) {
+    let result = bookings.filter((booking) => `${booking.firstName} ${booking.surname} ${booking.email}`.toLowerCase().includes(term.toLowerCase()))
+    result.length > 0 ? response.send(result) : response.send('Not found a booking with this term')
+  } else {
+    response.status(400).json('Bad request')
+  }
+});
+
+
+app.get("/bookings/:id", function (request, response) {
+  const id = Number(request.params.id)
+  if (!isNaN(id)) {
+    const booking = bookings.find(booking => booking.id === id)
+    if (booking) {
+      const booking = bookings.filter(booking => booking.id === id)
+      response.send(booking)
+    } else {
+      response.status(404).json({ msg: 'cannot be found by id' })
+    }
+  } else {
+    response.status(400).json({ msg: 'type a valid id' })
+  }
+});
+
+
+app.delete("/bookings/:id", function (request, response) {
+  const id = parseInt(request.params.id)
+  const booking = bookings.find(booking => booking.id === id)
+  if (booking) {
+    const currentBookings = bookings.filter(booking => booking.id !== id)
+    response.send({ msg: 'The booking is deleted', bookings: currentBookings })
+  } else {
+    response.status(404).json({ msg: 'cannot be found by id' })
+  }
+});
+
+app.post("/bookings", function (request, response) {
+  let keys = ["title", "firstName", "surname", "email", "roomId", "checkInDate", "checkOutDate"];
+  for (let key in request.body) {
+    if (!keys.includes(key)) {
+      response.status(400).json("We dont have such a data");
+      return;
+    } else if (!request.body[key] && keys.includes(key)) {
+      response.status(400).json("Please fill in all fields");
+      return;
+    }
+  }
+  if (request.body) {
+    moment(request.body.checkOutDate).isBefore(request.body.checkInDate) && response.status(400).json("CheckOutDate must be later then CheckInDate")
+    const emailRegex = /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/
```

Comment:
sounds good, all the hex code and weird chars 💯 
However, invalidation the principle of whitelisting is more prefered than black listing 
so easier email regex, including a-z, numbers and @ sign